jobs/base_model_validator.py

```diff
@@ -0,0 +1,192 @@
+            model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            beam.pvalue.TaggedOutput. An element of the output PCollection for
+            the doFn which represents an error as a key value pair.
+        """
+        element = self.clone_model(model)
+
+        if not self.regex.match(element.id):
+            yield errors.ModelInvalidIdError(element)
+
+
+class ValidateDeleted(BaseValidatorDoFn):
+    """DoFn to check whether models marked for deletion are stale."""
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            beam.pvalue.TaggedOutput. An element of the output PCollection for
+            the doFn which represents an error as a key value pair.
+        """
+        element = self.clone_model(model)
+        date_now = datetime.datetime.utcnow()
+
+        date_before_which_models_should_be_deleted = (
+            date_now -
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED
+        )
+
+        if element.last_updated < date_before_which_models_should_be_deleted:
+            yield errors.ModelExpiredError(element)
+
+
+class ValidateModelTimeFields(BaseValidatorDoFn):
+    """DoFn to check whether created_on and last_updated timestamps are
+    valid."""
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            beam.pvalue.TaggedOutput. An element of the output PCollection for
+            the doFn which represents an error as a key value pair.
+        """
+
+        element = self.clone_model(model)
+        if element.created_on > (element.last_updated + MAX_CLOCK_SKEW_SECS):
+            yield errors.ModelTimestampRelationshipError(element)
+
+        current_datetime = datetime.datetime.utcnow()
+        if (element.last_updated + MAX_CLOCK_SKEW_SECS) > current_datetime:
+            yield errors.ModelMutatedDuringJobError(element)
+
+
+class BaseModelValidator(beam.PTransform):
+    """Composite Beam Transform which returns a pipeline of validation
+    errors."""
+
+    def expand(self, model_pipe):
+        """Function that takes in a beam.PCollection of datastore models and
+        returns a beam.PCollection of validation errors.
+
+        Args:
+            model_pipe: beam.PCollection. A collection of models.
+
+        Returns:
+            beam.PCollection. A collection of errors represented as
+            key-value pairs.
+        """
+        not_deleted, deleted = (
+            model_pipe
+            | beam.Map(
+                lambda m: beam.pvalue.TaggedOutput(
+                    'deleted' if m.deleted else 'not_deleted', m))
+            .with_outputs('not_deleted', 'deleted'))
```

Comment:
Please keep the order of `deleted` and `not_deleted` consistent:

```suggestion
        deleted, not_deleted = (
            model_pipe
            | beam.Map(
                lambda m: beam.pvalue.TaggedOutput(
                    'deleted' if m.deleted else 'not_deleted', m))
            .with_outputs('deleted', 'not_deleted'))
```